### PR TITLE
Inherit zIndex from Group layer if present

### PR DESCRIPTION
**Pull request by Patchez**
Addresses [Issue-6654](https://github.com/openlayers/openlayers/issues/6654)

### CODE REVIEW

File: src/ol/layer/group.js [hunk 1]
```diff
@@ -221,6 +221,9 @@ ol.layer.Group.prototype.getLayerStatesArray = function(opt_states) {
         layerState.extent = ownLayerState.extent;
       }
     }
+    if (layerState.zIndex === 0 && ownLayerState.zIndex !== 0) {
```
Comment (gberaudo) [inline src/ol/layer/group.js]: @Patchez, the user may set a `zIndex` of 0 so it cannot be used to mean `no value`. Instead, the zIndex should be tested against `undefined`.

You will have to change the `zIndex` initialization at https://github.com/openlayers/openlayers/blob/master/src/ol/layer/base.js#L36 and change places where the zIndex is used to fallback to 0 if `undefined`.

Comment (Patchez) [inline src/ol/layer/group.js]: Thanks, didn't think of that.

Comment (Patchez) [inline src/ol/layer/group.js]: I've changed the `zIndex` initialization to allow `undefined` and changed the conditional:
```javascript
if (layerState.zIndex === undefined) {
      if (ownLayerState.zIndex !== undefined) {
        layerState.zIndex = ownLayerState.zIndex;
        layerState.layer.setZIndex(ownLayerState.zIndex);
      } else {
        layerState.zIndex = 0;
        ownLayerState.zIndex = 0;
        layerState.layer.setZIndex(0);
        ownLayerState.layer.setZIndex(0);
      }
    }
```
I also made changes to [`ol.renderer.Map.sortByZIndex`](https://github.com/openlayers/openlayers/blob/master/src/ol/renderer/map.js#L333) to order `zIndex: undefined` the same as `zIndex: 0`.  However, many existing tests expect a layer to initialize with a default `zIndex` of `0`.   Should I go ahead and change these tests to expect `zIndex: 0`?

Comment (gberaudo) [inline src/ol/layer/group.js]: Yes I think it makes sense.

**Comment by Patchez** [comment]
I think this now functions as expected and should be ready for review

File: test_rendering/spec/ol/map.test.js [hunk 2]
```diff
@@ -16,6 +16,7 @@ describe('ol.rendering.Map', function() {
     target = createMapDiv(50, 50);
 
     var vectorLayer = new ol.layer.Vector({
+      zIndex: 0,
```
Comment (gberaudo) [inline test_rendering/spec/ol/map.test.js]: Is this change necessary?

File: src/ol/layer/group.js [hunk 3]
```diff
@@ -221,6 +221,10 @@ ol.layer.Group.prototype.getLayerStatesArray = function(opt_states) {
         layerState.extent = ownLayerState.extent;
       }
     }
+    if (layerState.layer.getZIndex() === undefined) {
+      layerState.zIndex = ownLayerState.zIndex;
+      layerState.layer.setZIndex(ownLayerState.zIndex);
```
Comment (gberaudo) [inline src/ol/layer/group.js]: Do you need to set the zIndex here?
I would expect that the internal mechanism of computing default zIndex would not interfere with the values set by the user.

Comment (Patchez) [inline src/ol/layer/group.js]: I was debating this myself but figured since we are inheriting the `ol.layer.Group`'s zIndex we would want to carry that over to the properties.  I'll remove it if you don't think it makes sense.

File: test_rendering/spec/ol/map.test.js (hunk 2 above)
Comment (Patchez) [inline test_rendering/spec/ol/map.test.js]: I was having some weird behavior with the rendering tests, the generated image would visually match the expected image but get `Error: done() called multiple times`.  Adding `zIndex: 0` resolved these issues.

File: src/ol/layer/group.js (hunk 3 above)
Comment (gberaudo) [inline src/ol/layer/group.js]: The problem I see is that if the parent group zIndex change, the child zIndex will no longer be computed again, so yes I think it is necessary to change the above code.
I will be in vacations next week, so I will not be able to continue the review. But there are others in the community to continue the review. Maybe @fredj, or @ahocevar will be interested to help having your fix merged.

File: test_rendering/spec/ol/map.test.js (hunk 2 above)
Comment (Patchez) [inline test_rendering/spec/ol/map.test.js]: I dug into this issue a bit more and found the underlying cause.  I'll include the fix in the next commit.

Comment (Patchez) [inline test_rendering/spec/ol/map.test.js]: I addressed this in [Commit-b843d7d](https://github.com/openlayers/openlayers/pull/6661/commits/b843d7de52e4ea89746c05dbdef2b8f31ea57293)

File: src/ol/layer/group.js (hunk 3 above)
Comment (Patchez) [inline src/ol/layer/group.js]: I addressed this in [Commit-b843d7d](https://github.com/openlayers/openlayers/pull/6661/commits/b843d7de52e4ea89746c05dbdef2b8f31ea57293)

**Comment by Patchez** [comment]
@gberaudo How is this looking?  If if there are any more changes that are needed I'd be more than happy to make them

**Comment by Patchez** [comment]
@ahocevar I took a look at that build failure and while all of the tests passed, it still failed with this message:
```
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
```
Any ideas on what could be causing this?

**Comment by ahocevar** [comment]
@Patchez I restarted the build. The error you saw was probably because of a timeout of a service that was not available. However, your most recent commit does not address my review comments. Let me know if you have any questions.

**Comment by Patchez** [comment]
@ahocevar Sorry about that, could you clarify what you'd like to see?  I didn't fully understand what you were asking for.

**Comment by ahocevar** [comment]
@Patchez I'd prefer to see the zIndex **always** initialised with `undefined` instead of `0`. To make that work consistently, you can more or less go back to your original version of this pull request where it accepted `undefined`. You'll only need to modify `ol.renderer.Map.sortByZIndex` so it returns `0` when one or both compared `zIndex`s are `undefined`.

**Comment by Patchez** [comment]
Okay, thank you.  I'll get on that.

**Comment by Patchez** [comment]
Sorry about the delay, I've been very busy recently and haven't had time to work on this.  I'll hopefully be able work on this again soon.

**Comment by ahocevar** [comment]
@Patchez Are you planning to finish this?

**Comment by Patchez** [comment]
Sorry, @ahocevar I've been under some crazy deadlines extraneous to this.  I think I'll have a spare moment this week to take a look at wrapping this up but if someone wants to help out with this it would definitely be appreciated.

**Comment by Patchez** [comment]
@ahocevar I'm having trouble understanding what you mean by
> To make that work consistently, you can more or less go back to your original version of this pull request where it accepted undefined

Do you mean the first commit (`0ab0d659a8de613af8ac3609a5f6917d9e4eff55`)?  Because I don't think that one really changed how z-index is initialized.

Sorry again for the delays, I honestly got a little burned out there for a while but I am really trying to get back into this.
